File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/aws/ssm/AwsSsmParameters.java
package uk.nhs.nhsx.core.aws.ssm;

import com.amazonaws.services.simplesystemsmanagement.AWSSimpleSystemsManagement;
import com.amazonaws.services.simplesystemsmanagement.AWSSimpleSystemsManagementClientBuilder;
import com.amazonaws.services.simplesystemsmanagement.model.GetParameterRequest;
import com.amazonaws.services.simplesystemsmanagement.model.GetParameterResult;
import com.github.benmanes.caffeine.cache.Caffeine;
import com.github.benmanes.caffeine.cache.LoadingCache;

import java.time.Duration;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.function.Function;

import static java.lang.String.format;

public class AwsSsmParameters implements Parameters {

    private static final ExecutorService executor = Executors.newCachedThreadPool();

    private final AWSSimpleSystemsManagement ssmClient;
    private final Duration refreshInterval;

    public AwsSsmParameters(AWSSimpleSystemsManagement ssmClient, Duration refreshInterval) {
        this.ssmClient = ssmClient;
        this.refreshInterval = refreshInterval;
    }

    public AwsSsmParameters() {
        this(AWSSimpleSystemsManagementClientBuilder.defaultClient(), Duration.ofMinutes(2));
    }

    @Override
    public <T> Parameter<T> parameter(ParameterName name, Function<String, T> convert) {
        LoadingCache<ParameterName, T> loader = Caffeine.newBuilder()
            .executor(executor)
            .refreshAfterWrite(refreshInterval)
            .build(key -> getParameter(key, convert));

        return () -> {
            try {
                return loader.get(name);
            } catch (Exception e) {
                throw new RuntimeException(format("Unable to load parameter for %s", name), e);
            }
        };
    }

    private <T> T getParameter(ParameterName name, Function<String, T> convert) {
        GetParameterRequest request = new GetParameterRequest().withName(name.getValue());
        GetParameterResult result = ssmClient.getParameter(request);
        return convert.apply(result.getParameter().getValue());
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/aws/ssm/AwsSsmParameters.kt
package uk.nhs.nhsx.core.aws.ssm

import com.amazonaws.services.simplesystemsmanagement.AWSSimpleSystemsManagement
import com.amazonaws.services.simplesystemsmanagement.AWSSimpleSystemsManagementClientBuilder
import com.amazonaws.services.simplesystemsmanagement.model.GetParameterRequest
import com.github.benmanes.caffeine.cache.Caffeine
import java.time.Duration
import java.util.concurrent.Executors
import java.util.function.Function

class AwsSsmParameters @JvmOverloads constructor(
    private val ssmClient: AWSSimpleSystemsManagement = AWSSimpleSystemsManagementClientBuilder.defaultClient(),
    private val refreshInterval: Duration = Duration.ofMinutes(2)
) : Parameters {
    override fun <T> parameter(name: ParameterName, convert: Function<String, T>): Parameter<T> {
        val loader = Caffeine.newBuilder()
            .executor(executor)
            .refreshAfterWrite(refreshInterval)
            .build { key: ParameterName -> getParameter(key, convert) }

        return Parameter {
            try {
                loader[name] ?: error("No value loaded")
            } catch (e: Exception) {
                throw RuntimeException(String.format("Unable to load parameter for %s", name), e)
            }
        }
    }

    private fun <T> getParameter(name: ParameterName, convert: Function<String, T>): T {
        val request = GetParameterRequest().withName(name.value)
        val result = ssmClient.getParameter(request)
        return convert.apply(result.parameter.value)
    }

    companion object {
        private val executor = Executors.newCachedThreadPool()
    }
}
